jvm/src/test/scala/org/atnos/eff/ValidateEffectSpec.scala

```diff
@@ -105,15 +109,63 @@ class ValidateEffectSpec extends Specification with ScalaCheck { def is = s2"""
     val handle: E => Check[Unit] = { case e => tell[Comput, E](e).as(()) }
 
     val comp1: Check[Int] = for {
-      _ <- wrong[Comput, E]("1").catchWrong(handle)
-      _ <- wrong[Comput, E]("2").catchWrong(handle)
+      _ <- wrong[Comput, E]("1").catchFirstWrong(handle)
+      _ <- wrong[Comput, E]("2").catchFirstWrong(handle)
     } yield 0
 
     val comp2: Check[Int] = comp1
 
     comp2.runNel.runWriter.run ==== ((Right(0), List("1", "2")))
   }
 
+  def catchFirstWrongValue = {
+    val validate: Eff[S, String] =
+      for {
+        _ <- ValidateEffect.correct[S, String, Int](1)
+        _ <- ValidateEffect.wrong[S, String]("error1")
+        a <- EffMonad[S].pure(3)
+        _ <- ValidateEffect.wrong[S, String]("error2")
+      } yield a.toString
+
+    validate.catchFirstWrong((s: String) => pure(s)).runNel.run ==== Right("error1")
+  }
+
+  def catchAllWrongValuesLinear = {
+    val validate: Eff[S, String] =
+      for {
+        _ <- ValidateEffect.correct[S, String, Int](1)
+        _ <- ValidateEffect.wrong[S, String]("error1")
+        a <- EffMonad[S].pure(3)
+        _ <- ValidateEffect.wrong[S, String]("error2")
+      } yield a.toString
+
+    validate.catchAllWrongs((ss: NonEmptyList[String]) => pure(ss.mkString_("", ", ", ""))).runNel.run ==== Right("error1, error2")
+  }
+
+  def catchAllWrongValuesApplicative = {
```

Comment:
The name `Applicative` here might be a bit misleading. I was thinking this was about the `Applicative` instance of `Eff` not the `Applicative` instance of the container used to collect the values. Which makes me think that we should add some tests for `Eff` `Applicative` cases:
```
val action = ValidateEffect.wrong[S, String]("error1") |@| ValidateEffect.wrong[S, String]("error1")
// what does that return?
validate.catchAllWrongs . ...
```

Reply:
I agree, so, I've renamed this test, added three more and reorganized all a little bit to make everything more clean and to reduce duplications.

Reply:
Oh, and by the way, new applicative tests introduced one more warning which I do not have an idea how to fix (and as far as I know, it is a compiler bug which is fixed only in dotty). So, sorry.